beacon-chain/execution/engine_client.go

```diff
@@ -649,6 +651,10 @@ func (s *Service) retrievePayloadsFromExecutionHashes(
 		return nil, fmt.Errorf("could not fetch payload bodies by hash %#x: %v", executionHashes, err)
 	}
 
+	if len(payloadBodies) != len(executionHashes) {
```

Comment:
Should not this check be directly in the `GetPayloadBodiesByHash` function?

According to [the specification](https://github.com/ethereum/execution-apis/blob/main/src/engine/shanghai.md#engine_getpayloadbodiesbyhashv1), is such a length mismatch happen, it is a bug in the EL.

If this check is written directly in the `GetPayloadBodiesByHash` function, then all calls to this function will benefit of this check.

In the contrary, all callers of `GetPayloadBodiesByHash`will need to do the same check.

Reply:
In the case referenced in the issue description the node is optimistically syncing which indicates that the EL might have just come online, which is why they return nothing instead of empty payloads.  It might be a race condition on the EL's side but in the event of bugs across different clients we can just handle it on our end. I have pushed up the change to handle the mismatch in the function